File: App/Composition/UploadImageAttachments.m
//  UploadImageAttachments.m
//
//  Copyright 2014 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "UploadImageAttachments.h"
#import <ImgurAnonymousAPIClient/ImgurAnonymousAPIClient.h>
@import Photos;

#import "Awful-Swift.h"

@interface ImageTag : NSObject

@property (nonatomic) TextAttachment *attachment;
@property (nonatomic) NSRange range;
@property (nonatomic) CGSize imageSize;
@property (nonatomic) NSURL *URL;

@property (readonly, nonatomic) NSString *BBcode;

@end

@implementation ImageTag

- (NSString *)BBcode
{
    CGSize maxSize = [TextAttachment requiresThumbnailImageSize];
    BOOL requiresThumbnailing = self.imageSize.width > maxSize.width || self.imageSize.height > maxSize.height;
    NSString *t = requiresThumbnailing ? @"t" : @"";
    return [NSString stringWithFormat:@"[%@img]%@[/%@img]", t, self.URL.absoluteString, t];
}

@end

static NSArray * AttachmentsInString(NSAttributedString *string)
{
    NSMutableArray *attachments = [NSMutableArray new];
    [string enumerateAttribute:NSAttachmentAttributeName
                       inRange:NSMakeRange(0, string.length)
                       options:NSAttributedStringEnumerationLongestEffectiveRangeNotRequired
                    usingBlock:^(NSTextAttachment *attachment, NSRange range, BOOL *stop)
     {
         if ([attachment isKindOfClass:[TextAttachment class]]) {
             ImageTag *tag = [ImageTag new];
             tag.attachment = (TextAttachment *)attachment;
             tag.range = range;
             [attachments addObject:tag];
         }
     }];
    return attachments;
}

static NSArray * AssetURLsOrImagesForTags(NSArray *tags)
{
    NSMutableArray *images = [NSMutableArray new];
    for (ImageTag *tag in tags) {
        if (tag.attachment.assetURL) {
            // Images in the assets library can be uploaded directly from the library. Just need to grab the size of the image for timg purposes.
            PHAsset *asset = [PHAsset fetchAssetsWithALAssetURLs:@[tag.attachment.assetURL] options:nil].firstObject;
            tag.imageSize = CGSizeMake(asset.pixelWidth, asset.pixelHeight);
            [images addObject:tag.attachment.assetURL];
        } else {
            tag.imageSize = tag.attachment.image.size;
            [images addObject:tag.attachment.image];
        }
    }
    return images;
}

static NSProgress * UploadImages(NSArray *images, void (^completion)(NSArray *URLs, NSError *error)) {
    NSProgress *progress = [NSProgress progressWithTotalUnitCount:images.count];
    dispatch_group_t group = dispatch_group_create();
    NSPointerArray *URLs = [NSPointerArray strongObjectsPointerArray];
    URLs.count = images.count;
    for (id image in images) {
        dispatch_group_enter(group);
        [progress becomeCurrentWithPendingUnitCount:1];
        void (^uploadComplete)() = ^(NSURL *imgurURL, NSError *error) {
            if (error) {
                if (!progress.cancelled) {
                    [progress cancel];
                    if (completion) completion(nil, error);
                }
            } else {
                [URLs replacePointerAtIndex:[images indexOfObject:image] withPointer:(__bridge void *)imgurURL];
            }
            dispatch_group_leave(group);
        };
        if ([image isKindOfClass:[NSURL class]]) {
            [[ImgurAnonymousAPIClient sharedClient] uploadAssetWithURL:image filename:@"image.png" completionHandler:uploadComplete];
        } else {
            [[ImgurAnonymousAPIClient sharedClient] uploadImage:image withFilename:@"image.png" completionHandler:uploadComplete];
        }
        [progress resignCurrent];
    }
    dispatch_group_notify(group, dispatch_get_main_queue(), ^{
        if (progress.cancelled) {
            if (completion) completion(nil, [NSError errorWithDomain:NSCocoaErrorDomain code:NSUserCancelledError userInfo:nil]);
        } else {
            if (completion) completion(URLs.allObjects, nil);
        }
    });
    return progress;
}

static NSString * ReplaceAttachmentsWithTags(NSAttributedString *richText, NSArray *tags)
{
    NSMutableString *string = [NSMutableString stringWithString:richText.string];
    for (ImageTag *tag in tags.reverseObjectEnumerator) {
         [string replaceCharactersInRange:tag.range withString:tag.BBcode];
    }
    return string;
}

NSProgress * UploadImageAttachments(NSAttributedString *richText, void (^completion)(NSString *plainText, NSError *error))
{
    richText = [richText copy];
    NSProgress *progress = [NSProgress progressWithTotalUnitCount:1];
    dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
        NSArray *tags = AttachmentsInString(richText);
        if (tags.count == 0) {
            --progress.totalUnitCount;
            
            return dispatch_async(dispatch_get_main_queue(), ^{
                completion(richText.string, nil);
            });
        }
        
        if (progress.cancelled) {
            return dispatch_async(dispatch_get_main_queue(), ^{
                completion(nil, [NSError errorWithDomain:NSCocoaErrorDomain code:NSUserCancelledError userInfo:nil]);
            });
        }
        
        NSArray *assetsAndImages = AssetURLsOrImagesForTags(tags);
        [progress becomeCurrentWithPendingUnitCount:1];
        UploadImages(assetsAndImages, ^(NSArray *URLs, NSError *error) {
            if (error) {
                return completion(nil, error);
            }
            
            dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
                [URLs enumerateObjectsUsingBlock:^(NSURL *URL, NSUInteger i, BOOL *stop) {
                    ImageTag *tag = tags[i];
                    tag.URL = URL;
                }];
                NSString *plainText = ReplaceAttachmentsWithTags(richText, tags);
                dispatch_async(dispatch_get_main_queue(), ^{
                    completion(plainText, nil);
                });
            });
        });
        [progress resignCurrent];
    });
    return progress;
}

File: App/Composition/UploadImageAttachments.swift
//  UploadImageAttachments.swift
//
//  Copyright 2014 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import AwfulSettings
import Foundation
import ImgurAnonymousAPI
import os
import Photos
import UIKit

private let logger = Logger(subsystem: Bundle.main.bundleIdentifier!, category: "UploadImageAttachments")

/**
    Replaces image attachments in richText with [img] tags by uploading the images anonymously to Imgur.
 
    - parameter completion: A block to call on the main queue after replacement, which gets as arguments: the tagged string on success, or nil on failure; and an error on failure, or nil on success.
 
    - returns: A progress object that can cancel the image upload.
 */
func uploadImages(attachedTo richText: NSAttributedString, completion: @escaping (_ plainText: String?, _ error: Error?) -> Void) -> Progress {
    let progress = Progress(totalUnitCount: 1)
    
    let localCopy = richText.copy() as! NSAttributedString
    DispatchQueue.global(qos: DispatchQoS.QoSClass.default).async {
        let tags = localCopy.imageTags
        guard !tags.isEmpty else {
            progress.completedUnitCount += 1
            return DispatchQueue.main.async {
                completion(localCopy.string, nil)
            }
        }

        let localerCopy = localCopy.mutableCopy() as! NSMutableAttributedString
        let uploadProgress = uploadImages(fromSources: tags.map { $0.source }, completion: { (urls, error) in
            if let error = error {
                return DispatchQueue.main.async {
                    completion(nil, error)
                }
            }
            
            guard let urls = urls else { fatalError("no error should mean some URLs!") }
            for (url, tag) in zip(urls, tags).reversed() {
                localerCopy.replaceCharacters(in: tag.range, with: tag.BBcode(url as URL))
            }
            DispatchQueue.main.async {
                completion(localerCopy.string, nil)
            }
        })
        progress.addChild(uploadProgress, withPendingUnitCount: 1)
    }
    
    return progress
}

private func uploadImages(fromSources sources: [ImageTag.Source], completion: @escaping (_ urls: [URL]?, _ error: Error?) -> Void) -> Progress {
    let progress = Progress(totalUnitCount: Int64(sources.count))
    
    let group = DispatchGroup()
    
    var urls: [URL?] = Array(repeating: nil, count: sources.count)
    for (i, source) in sources.enumerated() {
        group.enter()
        
        func uploadComplete(_ url: URL?, error: Error?) {
            defer { group.leave() }
            
            if let error = error {
                if !progress.isCancelled { progress.cancel() }
                return DispatchQueue.main.async {
                    completion(nil, error)
                }
            }
            
            guard let url = url else { fatalError("no error should mean a URL!") }
            urls[i] = url
        }
        
        progress.becomeCurrent(withPendingUnitCount: 1)
        
        switch source {
        case .image(let image):
            ImgurUploader.shared.upload(image, completion: { result in
                switch result {
                case .success(let response):
                    uploadComplete(response.link, error: nil)
                case .failure(let error):
                    logger.error("Could not upload UIImage: \(error)")
                    uploadComplete(nil, error: error)
                }
            })
            
        case .photoAsset(let assetIdentifier):
            guard let asset = PHAsset.fetchAssets(withLocalIdentifiers: [assetIdentifier], options: nil).firstObject else {
                logger.error("Could not find asset corresponding to local identifier \(assetIdentifier)")
                uploadComplete(nil, error: ImageUploadError.missingIdentifiedAsset)
                break
            }
            
            ImgurUploader.shared.upload(asset, completion: { result in
                switch result {
                case .success(let response):
                    uploadComplete(response.link, error: nil)
                case .failure(let error):
                    logger.error("Could not upload PHAsset: \(error)")
                    uploadComplete(nil, error: error)
                }
            })
        }
        
        progress.resignCurrent()
    }
    
    group.notify(queue: DispatchQueue.global(qos: DispatchQoS.QoSClass.default)) { 
        if progress.isCancelled {
            completion(nil, NSError(domain: NSCocoaErrorDomain, code: NSUserCancelledError, userInfo: nil))
        } else {
            completion(urls.compactMap { $0 }, nil)
        }
    }
    
    return progress
}

enum ImageUploadError: Error {
    case missingIdentifiedAsset
}

private struct ImageTag {
    let range: NSRange
    let size: CGSize
    let source: Source
    
    enum Source {
        case image(UIImage)
        case photoAsset(String)
    }
    
    init(_ attachment: TextAttachment, range: NSRange) {
        self.range = range
        
        if
            let assetIdentifier = attachment.photoAssetIdentifier,
            let asset = PHAsset.fetchAssets(withLocalIdentifiers: [assetIdentifier], options: nil).firstObject
        {
            source = .photoAsset(assetIdentifier)
            size = CGSize(width: asset.pixelWidth, height: asset.pixelHeight)
        } else if let image = attachment.image {
            source = .image(image)
            size = image.size
        } else {
            fatalError("couldn't get image off of attachment")
        }
    }
    
    func BBcode(_ url: URL) -> String {
        let t: String
        if FoilDefaultStorage(Settings.automaticTimg).wrappedValue &&
            (size.width > TextAttachment.requiresThumbnailImageSize.width ||
             size.height > TextAttachment.requiresThumbnailImageSize.height)
        {
            t = "t"
        } else {
            t = ""
        }
        return "[\(t)img]\(url.absoluteString)[/\(t)img]"
    }
}

private extension NSAttributedString {
    var imageTags: [ImageTag] {
        var tags: [ImageTag] = []
        enumerateAttribute(.attachment, in: NSRange(location: 0, length: length), options: .longestEffectiveRangeNotRequired) { (attachment, range, stop) in
            guard let attachment = attachment as? TextAttachment else { return }
            tags.append(ImageTag(attachment, range: range))
        }
        return tags
    }
}
